// wwc-core/index.js
/**
 * Created by victor on 1/31/15.
 */

module.exports = function () {
    "use strict";

    /* Private Members */
    var communicationManager,
        faultManager,
        jobManager,
        metricsManager,
        fileSystemManager,
        schedulerManager,
        deploymentManager;

    /* Loading Modules */
    communicationManager = require("");
    faultManager = require("");
    jobManager = require("");
    metricsManager = require("");
    fileSystemManager = require("");
    schedulerManager = require("");
    deploymentManager = require("");

    /* Getters */
    this.getCommunicationManager = function () {
        return communicationManager;
    };

    this.getFaultManager = function () {
        return faultManager;
    };

    this.getJobManager = function () {
        return jobManager;
    };

    this.getMetricsManager = function () {
        return metricsManager;
    };

    this.getFileSystemManager = function () {
        return fileSystemManager;
    };

    this.getSchedulerManager = function () {
        return schedulerManager;
    };

    this.getDeploymentManager = function () {
        return deploymentManager;
    };


};
